psydac/api/tests/test_api_feec_3d.py

```diff
@@ -0,0 +1,206 @@
+
+    #==============================================================================
+    # Discrete objects: Psydac
+
+    domain_h = discretize(domain, ncells=ncells, comm=MPI.COMM_WORLD)
+    derham_h = discretize(derham, domain_h, degree=degree, periodic=periodic)
+
+    a1_h = discretize(a1, domain_h, (derham_h.V1, derham_h.V1), backend=PSYDAC_BACKEND_GPYCCEL)
+    a2_h = discretize(a2, domain_h, (derham_h.V2, derham_h.V2), backend=PSYDAC_BACKEND_GPYCCEL)
+
+    # StencilMatrix objects
+    M1 = a1_h.assemble().tosparse().tocsc()
+    M2 = a2_h.assemble().tosparse().tocsr()
+
+    # Diff operators
+    GRAD, CURL, DIV = derham_h.derivatives_as_matrices
+
+    # Porjectors
+    P0, P1, P2, P3  = derham_h.projectors(nquads=[5,5,5])
+
+    CURL     = CURL.tosparse().tocsr()
+
+    # initial conditions
+    e0_1 = lambda x, y, z: e_ex[0](0, x, y, z)
+    e0_2 = lambda x, y, z: e_ex[1](0, x, y, z)
+    e0_3 = lambda x, y, z: e_ex[2](0, x, y, z)
+
+    e0   = (e0_1, e0_2, e0_3)
+
+    b0_1 = lambda x, y, z : b_ex[0](0, x, y, z)
+    b0_2 = lambda x, y, z : b_ex[1](0, x, y, z)
+    b0_3 = lambda x, y, z : b_ex[2](0, x, y, z)
+
+    b0   = (b0_1, b0_2, b0_3)
+    
+    # project initial conditions
+    e0_log   = pull_1(e0, mapping)
+    e0_coeff = P1(e0_log).coeffs
+
+    b0_log   = pull_2(b0, mapping)
+    b0_coeff = P2(b0_log).coeffs
+
+    # time integrator
+    e_history, b_history = splitting_integrator(e0_coeff.toarray(), b0_coeff.toarray(), M1, M2, CURL, dt, niter)
+
+    # study of fields
+    b_history = [array_to_stencil(bi, derham_h.V2.vector_space) for bi in b_history]
+    b_fields  = [VectorFemField(derham_h.V2, bi).fields for bi in b_history]
+
+    bx_fields = [bi[0] for bi in b_fields]
+    by_fields = [bi[1] for bi in b_fields]
+    bz_fields = [bi[2] for bi in b_fields]
+
+    bx_value_fun = lambda x, y, z: evaluation_all_times(bx_fields, x, y, z)
+    by_value_fun = lambda x, y, z: evaluation_all_times(by_fields, x, y, z)
+    bz_value_fun = lambda x, y, z: evaluation_all_times(bz_fields, x, y, z)
+
+    x,y,z      = derham_h.V0.breaks
+    x, y       = 0.5, 0.5
+
+    b_values_0 = []
+    for zi in z:
+        b_value_phys  = push_2(bx_value_fun, by_value_fun, bz_value_fun, x, y, zi, mapping)
+        b_values_0.append(b_value_phys[0])
+    b_values_0  = np.array(b_values_0)
+
+    time_array  = np.linspace(0, dt*niter, niter + 1)
+    tt, zz      = np.meshgrid(time_array, z)
+
+    b_ex_values_0 = b_ex[0](tt, x, y, zz)
+
+    error = abs(b_values_0-b_ex_values_0).max()
+    return error
+
+###############################################################################
+#            SERIAL TESTS
+###############################################################################
+
+#==============================================================================
+# 3D Maxwell's equations with "Collela" map
+#==============================================================================
+def test_maxwell_3d_1():
```

Comment:
We need to add a non-periodic test. A standing wave in a cavity would work nicely...

Reply:
this will need more work in order to apply boundary conditions and it's not something we currently have with the scipy format

Reply:
OK. Let's not worry about it in this pull request.